ofs/apps/cpeng/readme.md

```diff
@@ -0,0 +1,88 @@
+# hps #
+
+## SYNOPSIS ##
+
+`hps OPTIONS SUBCOMMAND SUBCOMMAND_OPTIONS`
+
+## DESCRIPTION ##
+`hps` is an application to aid in the development, deployment, and debugging of
+an HPS (hard processor system) on an FPGA that is designed with one. The current
+version of the hps program assumes an AFU (accelerator functional unit) is
+configured into the FPGA that can be discovered/accessed via an OPAE library and
+used for communicating with the HPS. When invoked with one of its subcommands,
+hps will enumerate the AFU for that subcommand before executing it.
+
+
+## OPTIONS ##
+  -h,--help
+
+    Print this help message and exit
+
+  -p,--pci-address \<address\>
+
+    Use <address> in the filter for locating the AFU where <address> must be in
+    the following format: [<domain>:]<bus>:<device>.<function>
+
+  -l,--log-level \<level\>
+
+    stdout logging level. Must be one of:
+    {trace,debug,info,warning,error,critical,off}
+    Default is 'info'.
+
+  -s,--shared
+
+    open in shared mode, default is off
+  -t,--timeout \<timeout\>
+
+    Program timeout in milliseconds. Default is 60000.
+
+## SUBCOMMANDS ##
+  `cpeng`
+
+    Run copy engine command. The copy engine command is used to program copy
+    engine AFU registers to copy an image file from the host into the FPGA DDR.
+    The copy engine AFU is logic that can be used to copy an HPS image into the
+    FPGA DDR. When the HPS boots, the first stage boot loader will load an image
+    from a specific offset in DDR that will be used to transition into the second
+    stage boot loader and subsequently boot into the embedded Linux that is also
+    part of this image.
+
```

Comment:
The copy engine is a specific hardware implementation of something that essentially allows for downloading firmware to a SOC on the board.  This flow is very similar to the Partial Reconfiguration flow and the secure manager flow.  In all cases we are downloading some binary to be consumed by some portion of the card.

Since Partial Reconfiguration and secure downloads are both handled by the fpgasupdate command, shouldn't we put this functionality there as well?  The image being download is a FIT image.  The command, fpgasupdate, would have to notice this type and branch to code that talks to the copy engine.